x-pack/plugins/lens/public/app_plugin/app.tsx

```diff
@@ -309,88 +318,144 @@ export function App({
+      data.indexPatterns,
+      core.notifications
+    )
+      .then((indexPatterns) => {
+        // Don't overwrite any pinned filters
+        data.query.filterManager.setAppFilters(
+          injectFilterReferences(doc.state.filters, doc.references)
+        );
+        setState((s) => ({
+          ...s,
+          isLoading: false,
+          persistedDoc: doc,
+          lastKnownDoc: doc,
+          query: doc.state.query,
+          indexPatternsForTopNav: indexPatterns,
+        }));
+      })
+      .catch((e) => {
+        setState((s) => ({ ...s, isLoading: false }));
+
+        redirectTo();
+      });
+  };
+
   const runSave = async (
     saveProps: Omit<OnSaveProps, 'onTitleDuplicate' | 'newDescription'> & {
       returnToOrigin: boolean;
       onTitleDuplicate?: OnSaveProps['onTitleDuplicate'];
       newDescription?: string;
-    }
+    },
+    saveToLibrary: boolean = false
   ) => {
     if (!lastKnownDoc) {
       return;
     }
 
+    if (saveProps.newCopyOnSave && !state.byValueMode) {
+      if (embeddableEditorIncomingState?.embeddableId) {
+        embeddableEditorIncomingState.embeddableId = undefined;
+      }
+    }
+
     const doc = {
       ...getLastKnownDocWithoutPinnedFilters()!,
       description: saveProps.newDescription,
-      id: saveProps.newCopyOnSave ? undefined : lastKnownDoc.id,
+      savedObjectId: saveProps.newCopyOnSave ? undefined : lastKnownDoc.savedObjectId,
       title: saveProps.newTitle,
     };
 
-    await checkForDuplicateTitle(
-      {
-        ...doc,
-        copyOnSave: saveProps.newCopyOnSave,
-        lastSavedTitle: lastKnownDoc?.title,
-        getEsType: () => 'lens',
-        getDisplayName: () =>
-          i18n.translate('xpack.lens.app.saveModalType', {
-            defaultMessage: 'Lens visualization',
-          }),
-      },
-      saveProps.isTitleDuplicateConfirmed,
-      saveProps.onTitleDuplicate,
-      {
-        savedObjectsClient: core.savedObjects.client,
-        overlays: core.overlays,
-      }
-    );
-
-    const newlyCreated: boolean = saveProps.newCopyOnSave || !lastKnownDoc?.id;
-    docStorage
-      .save(doc)
-      .then(({ id }) => {
-        // Prevents unnecessary network request and disables save button
-        const newDoc = { ...doc, id };
-        const currentOriginatingApp = state.originatingApp;
-        setState((s) => ({
-          ...s,
-          isSaveModalVisible: false,
-          originatingApp:
-            newlyCreated && !saveProps.returnToOrigin ? undefined : currentOriginatingApp,
-          persistedDoc: newDoc,
-          lastKnownDoc: newDoc,
-        }));
-        if (docId !== id || saveProps.returnToOrigin) {
-          redirectTo(id, saveProps.returnToOrigin, newlyCreated);
+    const addedToLibrary = state.byValueMode && saveToLibrary;
+
+    const newlyCreated =
+      saveProps.newCopyOnSave || addedToLibrary || (!savedObjectId && !state.byValueMode);
```

Comment:
Please move this `const` inside the `docStorage.save()` promise, since that is the only usage.